### PR TITLE
Voxel ellipsoid math

### PR DESCRIPTION
**Pull request by jjhembd**
**Merge https://github.com/CesiumGS/cesium/pull/11847 first!**

# Description

Our preliminary implementation of the ellipsoid voxel shape depended on a few simplifying assumptions. Some of these assumptions broke down for non-spherical ellipsoids, introducing inaccuracies in places like:
- Calculation of geodetic latitude.
- Computing heights.

The errors were small on a global scale for nearly-spherical ellipsoids (like WGS84). But at smaller scales (<10km) the errors could be significant.

This PR improves the accuracy of ellipsoid math in the voxel shaders, with two key changes.

### Latitude surface for intersections
Latitude bounds are intersected as a conical surface. For a non-spherical ellipsoid, two things must be taken into account when constructing the cone:
1. The raymarching takes place in a warped coordinate space, where the ellipsoid has been stretched vertically to become a sphere. This warping must be reversed for latitude intersections. See `intersectFlippedCone` and `intersectRegularCone` in `IntersectEllipsoid.glsl`.
2. The apex of the cone is *not* at the center of the earth. It must be shifted along the z-axis by a latitude-dependent value. See `getLatitudeConeShift` in `IntersectEllipsoid.glsl`

### Reference ellipsoid for heights and latitudes
The previous code used the maximum height surface as the reference "ellipsoid" for intersections, but the minimum height surface as the reference for latitude and height calculations.

This PR uses the maximum height surface everywhere. See `convertUvToShapeUvSpace` in `convertUvToEllipsoid.glsl`. It also computes a more accurate normal for the latitude calculation, using the point on the ellipsoid surface below/above the current position. This point is available via the function `nearestPointOnEllipse` in `convertUvToEllipsoid.glsl`, which replaces the previous `ellipseDistanceIterative` method.

## Testing plan

Compare alignment of height and latitude bounds with sample boundaries in [this local Sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=7VoLb9s4Ev4rRLBA5FaVX0pix0nvUje9Buu0QZ3r9ZGgoSXa1i0leUnKtRP4v9/wpYftNO2hi7aHC4JIJGeGw2+GM0NGQZpwge6uEoT6mAnCI5y0Xdk8pTSa8TQKVeslwWGUTC4iEUzfpJSqznMsWLTw4X2FjlEfmLO4d5VcJYGSOo/IZ8JgJCGfzaj3VvU5VzuBavfTROAoIexqx9VajDAnA7wk7NCynMV4QthSdXpjlsYXLJ1HIWEnfJkEjmRClvYyouQcz4aEzaOAGE5Lr5j/yahhyZlGWUTD8zTMKFGjOyecE8Hrl2QhMkZ4/RWGJ6anAND07Oxqp6YFqEfNrWh9EQV/SN3HmHKihiYkDdKw2ieimFBYdtG3qpWAu0MBTuaYu4gHJCEuNGPCMJIoa1A1sSLyQNcTAYYYZYI4uwKPoiQki10X7TZ2az1Ur6OLTKBxGmQcpQkSU2LE5xLSJKCgN0gfZ0kgojRxatoahtJT3I7UcaWmtnQwPpMY9ZWCQyIEOAl3tJJGhl6TXkIfkIBpbnRLqi4dwlF0IeHAjaXYQ+0yuUM6v90ZjlnKI0nhLVYu2uxdbu29XWkrpSwi4G/FDOteXcwz1SMVcZKq3MGAQ0kG49307FJTSjyaTpxiwQo2iVoIKMaggYfD8HQOL4OICzAwc3b/IMsw/ZyA1XJoHWKNEI2h4QEJOj4+RrvBrh1AX8Zf6YRWhXd9I+q+3212fN9rdvaafsvvtl30ZL/TOmi19jy/4Xe63YOWi5rdbqfZbXl+p7nvdw46e52vRbvptZq+32z5+81mc6/rd0Hak6bXaMNUe3630/b9dudgz0V7XrPbaDb3D5r7jQO/1WnXii2j1+qpneJNaDoiHp+mn6Uzy73VyylgN5MkdMym151v0wWhZwmfkUCk7DxaREmttyYzJKNsMgSRLxjgxy8IGxIwdAgzCJaRyr6NoySKs/gNLDSD7RvjRdFU2zePqt6//jHs+D3LSmz/MMCU2JhZmKIiaU2wW50WPUINb18uI3cliIABCWUMG4IBKJFhUi3dhkaHT/HM+puYRhBVZAfooZ69vBtmepZmSchhaELEuW0aASVCvKgQ2uYGoWpfMpzwccpioDZZRUVrhYZTBafEG0IcTTiskG8i1nGR/C1RJ9J8QPgRsk9KU8g518WgWM70oPGPcyJwiAW+hH7v7WnfL9EGaTxLE3Duy3uY+mUC78Xg9cllu3Vtw4C2OA8Yho3Ql5pUc6TqcrR3P2g5b8ZSkUrtPUb+zGA3PwcFSqEcqXSXzuQ7hPxKUL6DTETJgMwJddXrO/14rx8flMsaVqWOjkU5E3p6jJp5MGIEQlGCwMpkDMktVICtNJuer2KvNQv2CjK5Gkug7PKxcV0aBl8JwTUIgS2oZ9bKPirJ95bV5gI91ivb3v2uJF4bUFtF9bAsEG/UnBJ9Ca9TiHCVtm5JqZpBysABhoojLq0DqWFOnI+5/Ott+fS/mdDUb8YHsZjKEK8FvMrikQxYJHSqGtrVzqUz9WFrClhuBZkvwXlbkhBMcZIUMrbtHwgA+ZZQdI5cQ+0e0OVWeEFTLNqtE8bw0inp+KgynYUaIgdyKBHoFtgbPXgcVbVFt48f536aUy819bJKvYSeEnWJfqHpF1X6BfRU6O2aKGEzwt4B0wLVkVMB9wnsm952hvfAsFxjWH6J4QMw3K4x3FqGgqVe10xTIJ9mBDzfKigTxh4w2Pllu52Pf1Dt1sbk05Kxpc8lkGDLTuds6iu3PeT3tak2RVMJtLe30R8YDykHSpUnXnLqTKFcdhEEMpjBrYTXKg5alCqSbfzQP7cPRpAtXYsNJfU+a307OpjOpljT6DQ49cY0hVRgJT6WoFQXk++bj3pB1f0hOa5VGJNAMROSv5KzWXBOGCHJt/C2Ct4Rzci3sLavtZPkPCv9skImnSBUiq65OJtct9Q85xkV0S9W8qzViR6D0i76lSufXMQnKisHmy32rd0eMtcPr3PWlf9fLHsk6n9h1SMFbS16GMGCXLAohvM6EM+MzV0UZFyA/ClWrRgO01TvFYN+5aA2swI4TBunc3ICZ0yrV6naySfKr1M2BcAJ3VwTlS+vKtxOnvCtvof5m2uHygs4rC7HkpRWdVhuGIKVvmZSlqqqDzkFA75iiOMZhdNBcSCVyX6N9t+REOoyzp6KN6RxQWbD6Jao3NvaShHIPDCc4YCcMqYycbtl8DVAzisHanXKPpdL8jaRr3RUvWdjrHRuGjM8kZcoGsVhNEnk7ZyUaAG9mUPczOnOcZQ4L0zjLJnJuzCuni4UASk0g9v4k8WdsAhTOF7rF4D+7qZ0wC/bb9vBrTSsvaOqrLxQPEQ3v93ds4aVXcGcBG0NwitIEFiWQ4l6AfM4Ul3dgk1tlqLh9XjGxmAczVQUGGNZTMP+H48zLoGCROU0ZJEUpsIpTeMqKGg0mYrnESNqc572a+uCFIH2N1k3qqpExRclPye2IHrFxFbb2GQLzxQmk1GvXgcRVvKmDFUb3S8B66B7s36TWjbYoFD7/zb7epvJdTmWvRBbr2+hLRb8pTXCUaUFh4DHWrQpab+/vYdFPPv+9taolkJmdcGCYagTnkdc4CSw8JZ4zCXfXiOvc80gDGCWn5IK2kcIj6CQNPNZwLaYIKeBQypIr1lk/6bRNmcjOYt95iSHmqQ6qo5TG/N9hXnKxhlCmRBgLiiRifUyTekIs3OSZM5HKdgkUqGgv9rJa19AoCgLkawernZMVgRrEwrefliq/0pHcO0NNhsbD3j4FrQ4ppQz/lCV67LyPB0Mzi6Gr8+eW8LCc+2MRYJbr2oK4RtVAtrMLOszrBSx/vv1eHEivhNk9xyifk3E+ksqT5/sJwSs/35w9ur56ZtfBK8fvCF/SrSepYufDqhnr9/9/Bj98M33g1G6rlXPxVtumzQC8nLC3FYdHz8UedcvKmwhlF8YqWunU0pktcN1kakuHy/OoFLbd9GTVkP91Mo3G1tuPVYbyq/dgP31yltsC/077mZfK+9rynV5japJHlihuf2Gkoauf90yLB2Q5f/3X2oip3LLoD+pkFMZGfK/NaqMOtHfXhQOLy8x5MIql1lKytrVRW/9Qwt5jWlmk83SeJxCoXhhPo2QBwszR/65RIl2Jr9mCcundnNRAv1ORVBt4/uHNV517ePeB5Qy+uD0xeWn/uCs/7sk3nF3jrhYUvLUWufvUTxLmUAZo47n1QWJZxT2Ia+PMphKeAHndk8d1cusR2E0R1F4vOWrIxRQzDmMjDNKZdl8tfP0qA70G6xQn8uvF17PCaN4KcmmzacD3el53lEdmts5ha531yT/Bw), which displays a highly-flattened ellipsoid to exaggerate the errors.

Before this PR:
image

After this PR:
image

# Author checklist

- [x] I have submitted a Contributor License Agreement
- [x] I have added my name to `CONTRIBUTORS.md`
- [x] I have updated `CHANGES.md` with a short summary of my change
- [ ] I have added or updated unit tests to ensure consistent code coverage
- [x] I have update the inline documentation, and included code examples where relevant
- [x] I have performed a self-review of my code

### CODE REVIEW

File: packages/engine/Source/Scene/VoxelEllipsoidShape.js [hunk 1]
```diff
@@ -166,100 +164,68 @@ VoxelEllipsoidShape.prototype.update = function (
...
-  const renderMinLongitude = Math.max(shapeMinLongitude, clipMinLongitude);
-  const renderMaxLongitude = Math.min(shapeMaxLongitude, clipMaxLongitude);
-
-  const shapeMinLatitude = CesiumMath.clamp(
-    minBounds.y,
-    defaultMinLatitude,
-    defaultMaxLatitude
+  // Don't let the height go below the center of the ellipsoid.
+  const radii = Matrix4.getScale(modelMatrix, scratchScale);
+  const actualMinBounds = Cartesian3.clone(DefaultMinBounds, new Cartesian3());
```
Comment (ggetz) [inline packages/engine/Source/Scene/VoxelEllipsoidShape.js:184]: Scratch variable for the result?

https://github.com/CesiumGS/cesium/blob/main/Documentation/Contributors/CodingGuide/README.md#result-parameters-and-scratch-variables

This also applies throughout this function.

**Review by ggetz — COMMENTED** [review]
Thanks @jjhembd! Code overall is looking good. Should we update `CHANGES.md` with a note that this should fix rendering bugs for ellipsoid shapes?

**Comment by jjhembd** [comment]
Updated, thanks @ggetz !

**Comment by ggetz** [comment]
Looks good to me! Thanks @jjhembd!

**Pull request merged**